File: ScaleformUI_Csharp/Menus/UIMenu/BreadcrumbsHandler.cs
namespace ScaleformUI.Menu
{
    internal static class BreadcrumbsHandler
    {
        private static readonly List<Tuple<UIMenu, dynamic>> breadcrumbs = new List<Tuple<UIMenu, dynamic>>();
        internal static int Count => breadcrumbs.Count;
        internal static int CurrentDepth => breadcrumbs.Count - 1;
        internal static UIMenu PreviousMenu => breadcrumbs[CurrentDepth - 1].Item1;
        public static bool SwitchInProgress = false;

        internal static void Forward(UIMenu menu, dynamic data)
        {
            breadcrumbs.Add(new Tuple<UIMenu, dynamic>(menu, data));
        }

        internal static void Clear()
        {
            breadcrumbs.Clear();
        }

        internal static void Backwards()
        {
            breadcrumbs.RemoveAt(CurrentDepth);
        }
    }
}

File: ScaleformUI_V8/src/menus/breadcrumbs-handler.ts
import { BaseMenu } from "./menu.base";

export class BreadcrumbsHandler {
    static breadcrumbs: [BaseMenu, any][] = [];
    public static SwitchInProgress: boolean = false

    static get Count(): number {
        return this.breadcrumbs.length;
    }

    static get CurrentDepth(): number {
        return this.breadcrumbs.length - 1;
    }

    public static  get PreviousMenu(): BaseMenu {
        return this.breadcrumbs[this.CurrentDepth - 1][0];
    }

    static Forward(menu: BaseMenu, data: any) {
        this.breadcrumbs.push([menu, data]);
    }

    static Clear() {
        this.breadcrumbs.length = 0;
    }

    static Backwards() {
        this.breadcrumbs.pop();
    }
}